### ISSUE TITLE
Pathtools: Multiple changes to $ENV{TMPDIR} are not reflected by File::Spec->tmpdir()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11308]
Migrated from [rt.perl.org#89940](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=89940) (status was 'resolved')

Searchable as RT89940$

**Comment by p5pRT** [comment]
### From nickl@ebi.ac.uk

#### Created by nickl@ebi.ac.uk



Multiple changes to $ENV\{TMPDIR\} are not reflected by File&#8203;::Spec\->tmpdir\(\) because it caches the result on the first call\.
This is a problem if the value of $ENV\{TMPDIR\} changes during the lifetime of the script \- especially problematic for scripts running in a persistent environment such as mod\_perl\.

The tmpdir function is implemented by platform specific modules\, e\.g\. File&#8203;::Spec&#8203;::Unix\, File&#8203;::Spec&#8203;::Win32 etc\.

Here are the steps to repeat the issue under Unix&#8203;:

1\. Create some tmp dirs to test with&#8203;:

&nbsp;  $ mkdir /tmp/foo
&nbsp;  $ mkdir /tmp/bar

2\. Create and run the following Perl script&#8203;:

&nbsp;  use File&#8203;::Spec;

&nbsp;  $ENV\{TMPDIR\} = '/tmp/foo';
&nbsp;  print File&#8203;::Spec\->tmpdir\(\) \. "\\n";

&nbsp;  $ENV\{TMPDIR\} = '/tmp/bar';
&nbsp;  print File&#8203;::Spec\->tmpdir\(\) \. "\\n";


The script output will be&#8203;:
/tmp/foo
/tmp/foo

The second call to File&#8203;::Spec\->tmpdir\(\) has not detected the change in value of $ENV\{TMPDIR\}\.

The modules for platforms other than Unix follow the same style of implementation which suggests to me that these will all have the same problem but I haven't tested them\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.8 in the Red Hat build system.
It is being executed now by Perl v5.8.8 - Thu Sep  3 10:25:37 EDT 2009.

Site configuration information for perl v5.8.8:

Configured by Red Hat, Inc. at Thu Sep  3 10:25:37 EDT 2009.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.18-53.el5, archname=x86_64-linux-thread-multi
    uname='linux builder10.centos.org 2.6.18-53.el5 #1 smp mon nov 12 02:14:55 est 2007 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Doptimize=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic -Dversion=5.8.8 -Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Dlibpth=/usr/local/lib64 /lib64 /usr/lib64 -Dprivlib=/usr/lib/perl5/5.8.8 -Dsitelib=/usr/lib/perl5/site_perl/5.8.8 -Dvendorlib=/usr/lib/perl5/vendor_perl/5.8.8 -Darchlib=/usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi -Dsitearch=/usr/lib64/perl5/site_perl/5.8.8/x86_64-linux-thread-multi -Dvendorarch=/usr/lib64/perl5/vendor_perl/5.8.8/x86_64-linux-thread-multi -Darchname=x86_64-linux-thread-multi -Dvendorprefix=/usr -Dsiteprefix=/usr -Duseshrplib -Dusethreads -Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db -Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio -Dinstallusrbinperl=n -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less -isr -Dd_gethostent_r_proto -Ud_en!
 dhostent_r_proto -Ud_sethostent_r_proto -Ud_endprotoent_r_proto -Ud_setprotoent_r_proto -Ud_endservent_r_proto -Ud_setservent_r_proto -Dinc_version_list=5.8.7 5.8.6 5.8.5 -Dscriptdir=/usr/bin'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=define uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -Wdeclaration-after-statement -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='4.1.2 20080704 (Red Hat 4.1.2-46)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =''
    libpth=/usr/local/lib64 /lib64 /usr/lib64
    libs=-lresolv -lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lresolv -lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic'

Locally applied patches:
    


@INC for perl v5.8.8:
    /homes/nickl/lib/perl5/x86_64-linux-thread-multi
    /homes/nickl/lib/perl5
    /homes/nickl/perl5/lib/perl5//x86_64-linux-thread-multi
    /homes/nickl/perl5/lib/perl5/
    /nfs/ensembl/perllib2/x86_64-linux-thread-multi
    /nfs/ensembl/perllib2
    /nfs/ensembl/bioperl-live/
    /usr/lib64/perl5/site_perl/5.8.8/x86_64-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.8
    /usr/lib/perl5/site_perl
    /usr/lib64/perl5/vendor_perl/5.8.8/x86_64-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.8
    /usr/lib/perl5/vendor_perl
    /usr/lib64/perl5/5.8.8/x86_64-linux-thread-multi
    /usr/lib/perl5/5.8.8
    .


Environment for perl v5.8.8:
    HOME=/homes/nickl
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/ebi/lsf/extsrv/6.2/linux2.6-glibc2.3-x86_64/lib:/nfs/ensembl/xerces-c-3.1.1/lib
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/ebi/lsf/extsrv/6.2/linux2.6-glibc2.3-x86_64/bin:/ebi/lsf/extsrv/6.2/linux2.6-glibc2.3-x86_64/etc:/homes/nickl/bin:/nfs/ensembl/bin:/homes/nickl/bin
    PERL5LIB=:/homes/nickl/lib/perl5:/homes/nickl/perl5/lib/perl5/:/nfs/ensembl/perllib2:/nfs/ensembl/bioperl-live/
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Thu May 05 09&#8203;:15&#8203;:22 2011\, nickl@&#8203;ebi\.ac\.uk wrote&#8203;:
> Multiple changes to $ENV\{TMPDIR\} are not reflected by File&#8203;::Spec\-
>    >tmpdir\(\) because it caches the result on the first call\.
> This is a problem if the value of $ENV\{TMPDIR\} changes during the
>    lifetime of the script \- especially problematic for scripts running
>    in a persistent environment such as mod\_perl\.
> 

Confirmed \-\- but could it be argued that if a program is to be run in a
persistent environment\, and then you change that environment in
mid\-course \(which is what redefining %ENV entails\)\, you get what you
deserve?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Nov 25 17&#8203;:16&#8203;:21 2011\, jkeenan wrote&#8203;:
> On Thu May 05 09&#8203;:15&#8203;:22 2011\, nickl@&#8203;ebi\.ac\.uk wrote&#8203;:
> > Multiple changes to $ENV\{TMPDIR\} are not reflected by File&#8203;::Spec\-
> >    >tmpdir\(\) because it caches the result on the first call\.
> > This is a problem if the value of $ENV\{TMPDIR\} changes during the
> >    lifetime of the script \- especially problematic for scripts running
> >    in a persistent environment such as mod\_perl\.
> > 
> 
> Confirmed \-\- but could it be argued that if a program is to be run in a
> persistent environment\, and then you change that environment in
> mid\-course \(which is what redefining %ENV entails\)\, you get what you
> deserve?
> 

You could\, but then one could argue back that modifying env vars
throughout a program is common practice\.  \(At least I had to write ten
lines to juggle various env localisations recently\, because LWP and
Net&#8203;::SSL were refusing to play along nicely without that\.\)  I think I
agree with the original report\.  A module shouldn’t change behaviour
\(stop reading %ENV\) depending on whether you’ve used a function in it
before\.


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @tsee

On 11/26/2011 07&#8203;:19 AM\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Nov 25 17&#8203;:16&#8203;:21 2011\, jkeenan wrote&#8203;:
>> On Thu May 05 09&#8203;:15&#8203;:22 2011\, nickl@&#8203;ebi\.ac\.uk wrote&#8203;:
>>> Multiple changes to $ENV\{TMPDIR\} are not reflected by File&#8203;::Spec\-
>>>     >tmpdir\(\) because it caches the result on the first call\.
>>> This is a problem if the value of $ENV\{TMPDIR\} changes during the
>>>     lifetime of the script \- especially problematic for scripts running
>>>     in a persistent environment such as mod\_perl\.
>>>
>>
>> Confirmed \-\- but could it be argued that if a program is to be run in a
>> persistent environment\, and then you change that environment in
>> mid\-course \(which is what redefining %ENV entails\)\, you get what you
>> deserve?
>>
>
> You could\, but then one could argue back that modifying env vars
> throughout a program is common practice\.  \(At least I had to write ten
> lines to juggle various env localisations recently\, because LWP and
> Net&#8203;::SSL were refusing to play along nicely without that\.\)  I think I
> agree with the original report\.  A module shouldn’t change behaviour
> \(stop reading %ENV\) depending on whether you’ve used a function in it
> before\.

Not that I disagree\, but at the same time\, we've seen many people 
complain about the dismal performance of some File&#8203;::Spec operations\. 
Maybe this is a case where attempting to cache things for performance 
went wrong? Maybe it even matters\.

Just something for consideration\. File&#8203;::Spec could do with some love 
that I can't give it\.

Cheers\,
Steffen

**Comment by p5pRT** [comment]
### From nickl@ebi.ac.uk


On 26/11/2011 06&#8203;:19\, Father Chrysostomos via RT wrote&#8203;:
> On Fri Nov 25 17&#8203;:16&#8203;:21 2011\, jkeenan wrote&#8203;:
>    
>> On Thu May 05 09&#8203;:15&#8203;:22 2011\, nickl@&#8203;ebi\.ac\.uk wrote&#8203;:
>>      
>>> Multiple changes to $ENV\{TMPDIR\} are not reflected by File&#8203;::Spec\-
>>>     >tmpdir\(\) because it caches the result on the first call\.
>>> This is a problem if the value of $ENV\{TMPDIR\} changes during the
>>>     lifetime of the script \- especially problematic for scripts running
>>>     in a persistent environment such as mod\_perl\.
>>>
>>>        
>> Confirmed \-\- but could it be argued that if a program is to be run in a
>> persistent environment\, and then you change that environment in
>> mid\-course \(which is what redefining %ENV entails\)\, you get what you
>> deserve?
>>
>>      
> You could\, but then one could argue back that modifying env vars
> throughout a program is common practice\.  \(At least I had to write ten
> lines to juggle various env localisations recently\, because LWP and
> Net&#8203;::SSL were refusing to play along nicely without that\.\)  I think I
> agree with the original report\.  A module shouldn’t change behaviour
> \(stop reading %ENV\) depending on whether you’ve used a function in it
> before\.
>    

I would argue that the caching should be removed\. Failing that\, at the 
very least the File&#8203;::Spec documentation should contain a warning that 
the first tmpdir value is cached\, otherwise confusion will continue to 
arise\.

An example&#8203;: Bio&#8203;::DB&#8203;::Sam uses File&#8203;::Spec to manage it's temp dir\. It 
makes the assumption that the location can be changed by changing the 
environment\. Here's a snippet from the Bio&#8203;::DB&#8203;::Sam docs&#8203;:

&nbsp;   When opening a remote BAM file\, you may not wish for the index to be 
downloaded to the
&nbsp;   current working directory\. This version of index\_open copies the 
index into the directory
&nbsp;   indicated by the TMPDIR environment  variable or the system\-defined 
/tmp directory if not
&nbsp;   present\. You may change the environment variable just before the call 
to change its behavior\.

This is in fact incorrect \- but it seems a resaonlable assumption for 
the author to have made given the current File&#8203;::Spec tmpdir documentation\.

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Nov 27 04&#8203;:33&#8203;:08 2011\, smueller@&#8203;cpan\.org wrote&#8203;:
> On 11/26/2011 07&#8203;:19 AM\, Father Chrysostomos via RT wrote&#8203;:
> > On Fri Nov 25 17&#8203;:16&#8203;:21 2011\, jkeenan wrote&#8203;:
> >> On Thu May 05 09&#8203;:15&#8203;:22 2011\, nickl@&#8203;ebi\.ac\.uk wrote&#8203;:
> >>> Multiple changes to $ENV\{TMPDIR\} are not reflected by File&#8203;::Spec\-
> >>>     >tmpdir\(\) because it caches the result on the first call\.
> >>> This is a problem if the value of $ENV\{TMPDIR\} changes during the
> >>>     lifetime of the script \- especially problematic for scripts
running
> >>>     in a persistent environment such as mod\_perl\.
> >>>
> >>
> >> Confirmed \-\- but could it be argued that if a program is to be run in a
> >> persistent environment\, and then you change that environment in
> >> mid\-course \(which is what redefining %ENV entails\)\, you get what you
> >> deserve?
> >>
> >
> > You could\, but then one could argue back that modifying env vars
> > throughout a program is common practice\.  \(At least I had to write ten
> > lines to juggle various env localisations recently\, because LWP and
> > Net&#8203;::SSL were refusing to play along nicely without that\.\)  I think I
> > agree with the original report\.  A module shouldn’t change behaviour
> > \(stop reading %ENV\) depending on whether you’ve used a function in it
> > before\.
> 
> Not that I disagree\, but at the same time\, we've seen many people 
> complain about the dismal performance of some File&#8203;::Spec operations\. 
> Maybe this is a case where attempting to cache things for performance 
> went wrong? Maybe it even matters\.
> 
> Just something for consideration\. File&#8203;::Spec could do with some love 
> that I can't give it\.

It’s possible to have the best of both worlds&#8203;: cache the tmpdir\, but
check whether the environment has changed before returning it again\.

I have implemented this in commit 82730d4\.

\(I only thought about smoke\-me after pushing this to blead\.  I don’t
think it’s worth a revert because of that\.  We’ll see what happens\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
